docs/src/appendix/packaging.md

```diff
@@ -0,0 +1,501 @@
+
+The contents of `component.wasm` here is not an actual Wasm module, but simply the binary representation
+of the WIT package.
+
+## Directory Layout
+
+The canonical layout of a package when extracted as a directory, is as follows:
+
+    name@version/
+    |- manifest.toml
+    |- component.wasm
+    |- lib/
+       |- 0x00000000.mast
+       |- 0x00000001.mast
+    |- rodata/
+       |- 0x0.bin
+    |- etc/
+       |- component.sourcemap
+       |- docs/
+          |- index.html
+       |- vendor/
+          |- foo@0.1.0.mpkg
+
+### Manifest
+
+The `manifest.toml` file contains all of the top-level metadata about the package. When produced as a result
+of extracting a package, it's contents may differ from the `manifest.toml` used to create the package. This
+is because we provide some helpful functionality in the manifest for creating packages by hand, that are
+translated into canonical form when the package is created. These "canonicalizations" are covered below
+in each relevant section.
+
+> [!NOTE]
+> TOML was chosen as the format for the manifest due to it being a nicer format for
+> writing by hand, while being strongly typed, unambiguous, and supported by good
+> tooling in most languages.
+
+The manifest begins with basic metadata about the package:
+
+```toml
+[package]
+name = "my-package"
+version = "0.1.0"
+miden-version = "0.8.0"  # Optional
+```
+
+The `miden-version` field lets you specify a minimum Miden VM version for the package. Any attempt to load
+a package with a lower-versioned VM will fail. This field is optional, and when not present, is equivalent
+to no minimum required version.
+
+#### Package Type
+
+There are two primary types of packages: libraries and programs (executables). These correspond to the
+`[lib]` and `[bin]` sections of the manifest, respectively. Only one section is permitted.
+
+For libraries:
+
+```toml
+[lib]
+start = "0x..." # Optional
+```
+
+The `start` field specifies the MAST root of a function that must be called to initialize the library before
+any of its other exports are called. This function must be called for each new context in which the library
+is used, and must only be called once.
+
+For programs:
+
+```toml
+[bin]
+entrypoint = "0x..."
+```
+
+The `entrypoint` field specifies the MAST root of a function that will be called as the entrypoint of the
+program, e.g. `_start`. An entrypoint is required, and the root specified _must_ be contained within the
+primary MAST of the package (i.e. not in a dependency).
+
+#### Read-Only Data/Data Segments
+
+If the package has static data it expects to be present in linear memory when any of the code it contains starts
+executing, it should be specified using the `[rodata]` section:
+
+```toml
+[rodata]
+# Write the contents of `assets/foo.json` into linear memory at address 0x0,
+0x0 = { path = "assets/foo.json", encoding = "bytes" }
+0xDEADBEEF = { content = "0x...", encoding = "words" }
```

Comment:
I guess what I had in mind here was that a hand-maintained package could use a human-friendly layout and filenames, and yet be easily encoded to the canonical form, while retaining the ability to define data in both byte-addressable and word-addressable forms. Separating the metadata and the content is also more explicit, so we avoid any ambiguities.

To clarify what I mean, take for example a file named `0x0.json` - it doesn't tell you what is in that file just the fact that it is JSON (and JSON has no support for comments, so you can't put a comment inside the file to describe what it is, though that's a detail specific to JSON in this case). It's also non-obvious that the filename is the address at which that data will be written. Further, `.bin` is not an uncommon file extension for binary data, so it could very well be that someone has a `.bin` file they want to store in `rodata`, but don't realize that this will unintentionally be treated as already in canonical form. Obviously we could use a different file extension, but I think the general point I'm trying to make is just that there is a lot of implicit magic happening in that scenario that the manifest form makes explicit and non-magical.

That's not to say that we couldn't do it that way, but I do think there is value in being able to read the manifest and immediately be able to understand what the contents of `rodata` are, and how they will be processed by the package tooling.

Since both you and @greenhat had similar questions, perhaps that is evidence that it is in fact useful/important to make it possible to extract a package in the form it was originally defined, i.e. encoding a package and then extracting it with an option like `--no-rodata-canonicalization` would give you back the exact same contents on disk as were used to create the package. I had specified in my initial draft that we would always encode the canonicalized form, since that's all the VM needs, and so we would drop the `[rodata]` metadata we don't need in the process, like the filename and encoding type, but there is no reason why we can't support both. Doing so requires preserving that metadata in the package so that we can recover the original filenames and contents, but there are no technical limitations.